rabbitmq/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/RabbitMQSink.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2013-2022 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.collectors.scalastream
+package sinks
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
+
+import cats.syntax.either._
+
+import com.rabbitmq.client.Channel
+
+import com.snowplowanalytics.snowplow.collectors.scalastream.model.RabbitMQBackoffPolicyConfig
+
+class RabbitMQSink(
+  channel: Channel,
+  exchangeName: String,
+  backoffPolicy: RabbitMQBackoffPolicyConfig
+) extends Sink {
+
+  implicit val ec = ExecutionContext.global
+
+  override val MaxBytes = 128000000
+
+  override def storeRawEvents(events: List[Array[Byte]], key: String): Unit =
+    if (events.nonEmpty) {
+      log.info(
+        s"Sending ${events.size} Thrift records to exchange $exchangeName"
+      )
+      Future.sequence(events.map(e => sendOneEvent(e))).onComplete {
+        case Success(_) =>
+          log.debug(
+            s"${events.size} events successfully sent to exchange $exchangeName"
+          )
+        // We should never reach this as the writing of each individual event is retried forever
+        case Failure(e) =>
+          throw new RuntimeException(s"Error happened during the sending of ${events.size} events: ${e.getMessage}")
+      }
+    }
+
+  private def sendOneEvent(bytes: Array[Byte], currentBackOff: Option[FiniteDuration] = None): Future[Unit] =
+    Future {
```

Comment:
I think this is where it manifests the most. If I recall correctly this will be running on main execution context's thread pool (not global, but akka actor system's top-level). The Future will block the thread it's executed on. Might not be of a huge impact but once AMQP goes down this might saturate the top-level thread pool with retries.
Another way to do this would be to make use of `akka.ActorSystem`'s  `scheduler.schedule*` method.

Reply:
I considered using `akka.ActorSystem`'s `scheduler.schedule*` but at the time when we create the sink, `ActorSystem` does not exists yet, and it would require some refactor in the design of the collector, so I didn't go this way